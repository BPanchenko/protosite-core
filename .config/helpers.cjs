const path = require('path');
const { kebabCase, trim } = require('lodash');

const ROOT = process.cwd();
const OUTPUT = path.resolve(ROOT, `esm`);

const getFileEntryByRelativePath = (relativePath, typeName = '') => {
  const absolutePath = path.resolve(ROOT, relativePath);
  const props = path.parse(absolutePath);

  let name = kebabCase(props.name + '-' + (typeName || ''));
  name = trim(name, '-');

  return [name, absolutePath];
};

const types = Object.create(
  {
    get selected() {
      const proto = Object.getPrototypeOf(this);
      const key = proto.selectedKey;
      return key ? this[key] : null;
    },
    selectedKey: false,
    selectByProperty: (name, value) => {
      const proto = Object.getPrototypeOf(types);
      const keys = Object.keys(types);
      for (const key of keys) {
        const data = types[key];
        const selected = data.hasOwnProperty(name) && data[name] === value;
        proto.selectedKey = selected ? key : false;
        if (selected) break;
      }
      return proto.selectedKey;
    }
  },
  {
    CE: {
      value: Object.freeze({
        name: 'CustomElement',
        dashed: 'custom-element',
        plural: 'custom-elements',
        source: 'component'
      }),
      enumerable: true
    },
    CP: {
      value: Object.freeze({
        name: 'CorePart',
        dashed: 'core-part',
        plural: 'core-parts',
        source: 'trunk'
      }),
      enumerable: true
    },
    WC: {
      value: Object.freeze({
        name: 'WebComponent',
        dashed: 'web-component',
        plural: 'web-components',
        source: 'component'
      }),
      enumerable: true
    }
  }
);

module.exports = {
  ROOT,
  OUTPUT,
  getFileEntryByRelativePath,
  types
};
