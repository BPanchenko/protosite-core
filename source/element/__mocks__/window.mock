Object.defineProperty(window.console, 'assert', {
	writable: true,
	value: jest.fn().mockImplementation((assertion, message) => {
		if (assertion === false) {
			throw new Error(message)
		}
	}),
})
