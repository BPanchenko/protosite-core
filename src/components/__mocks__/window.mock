Object.defineProperty(window.console, 'assert', {
	writable: true,
	value: jest.fn().mockImplementation((assertion, message) => {
		if (assertion === false) {
			throw new Error(message)
		}
	}),
})

Object.defineProperty(window.document, 'fonts', {
	writable: true,
	value: {
		values: () => [
			{
				fontFace: {
					family: 'IconFont'
				},
			},
		]
	},
})
