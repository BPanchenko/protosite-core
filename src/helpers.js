const toString = Object.prototype.toString

Object.defineProperty(String.prototype, 'capitalize', {
	value: function () {
		return this.charAt(0).toUpperCase() + this.slice(1)
	},
	enumerable: false,
})

export const compactObject = (obj) => {
	const result = Array.from(Object.entries(obj)).reduce(
		(acc, [key, value]) => (value ? (acc[key] = value) && acc : acc),
		{},
	)
	return result
}

export function createElement(tagName, attrs = {}) {
	const element = document.createElement(tagName)
	Object.entries(attrs).forEach(([key, value]) => {
		switch (key) {
			case 'className':
				element.classList.add(value)
				break
			case 'classNames':
				value.forEach((item) => element.classList.add(item))
				break
			default:
				element.setAttribute(key, value)
		}
	})
	return element
}

export function createGridObject(step = 10) {
	const grid = document.createElement('o-grid')
	if (isNumber(step) && step > 0) grid.dataset.step = step

	const fragment = document.createDocumentFragment()
	fragment.appendChild(grid)

	return fragment
}

export function getTag(value) {
	if (value == null) {
		return value === undefined ? '[object Undefined]' : '[object Null]'
	}
	return toString.call(value)
}

export function isNumber(value) {
	return (
		typeof value === 'number' ||
		(isObjectLike(value) && getTag(value) === '[object Number]')
	)
}

export function isObjectLike(value) {
	return typeof value === 'object' && value !== null
}

export function selectNode(node) {
	if (document.body.createTextRange) {
		const range = document.body.createTextRange()
		range.moveToElementText(node)
		range.select()
	} else if (window.getSelection) {
		const selection = window.getSelection()
		const range = document.createRange()
		range.selectNodeContents(node)
		selection.removeAllRanges()
		selection.addRange(range)
	} else {
		console.warn('Could not select text in node: Unsupported browser.')
	}
}
