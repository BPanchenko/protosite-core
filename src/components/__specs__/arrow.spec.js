import '../__mocks__/window.mock'
import { compactObject } from '../../helpers.js'
import ArrowComponent from '../arrow.js'

describe('[ArrowComponent]', () => {
	describe('Attributes:', () => {
		describe.each`
			glyph           | weight     | direction        | figure    | style
			${'right-line'} | ${null}    | ${null}          | ${null}   | ${null}
			${null}         | ${'thick'} | ${'bottom'}      | ${null}   | ${'fill-angled'}
			${'left-line'}  | ${null}    | ${'right'}       | ${'line'} | ${'oblique'}
			${null}         | ${null}    | ${'bottom-left'} | ${'fill'} | ${'large-acute'}
		`(
			'[Glyph = "$glyph"; Weight = "$weight"; Direction = "$direction"; Figure = "$figure"; Style = "$style"]',
			(options) => {
				const avatar = new ArrowComponent(compactObject(options))
				document.body.appendChild(avatar)

				it('Light DOM:', () => expect(avatar).toMatchSnapshot())
				it('Shadow DOM:', () =>
					expect(avatar.shadowRoot.children).toMatchSnapshot())
			},
		)

		describe.each`
			glyph                                          | size
			${'right-line'}                                | ${null}
			${'right-angle-top-right-fill-acute'}          | ${'xxs'}
			${'right-angle-thick-right-top-fill-acute'}    | ${'xs'}
			${'right-angle-bottom-right-fill-angled'}      | ${'sm'}
			${'right-angle-thick-bottom-left-fill-angled'} | ${'md'}
			${'right-angle-right-bottom-fill'}             | ${'lg'}
			${'right-angle-thick-top-right-fill'}          | ${'xl'}
			${'right-angle-bottom-right-fill'}             | ${'xxl'}
			${'right-angle-right-top-line'}                | ${null}
			${'right-angle-right-bottom-fill-acute'}       | ${'another-size'}
		`('[Glyph = "$glyph"; Size = "$size"]', ({ glyph, size }) => {
			const avatar = new ArrowComponent({ glyph })
			document.body.appendChild(avatar)

			try {
				avatar.size = size

				it('Light DOM:', () => expect(avatar).toMatchSnapshot())
				it('Shadow DOM:', () =>
					expect(avatar.shadowRoot.children).toMatchSnapshot())
			} catch (error) {
				it('Error:', () => expect(error).toMatchSnapshot())
			}
		})

		describe.each`
			font
			${null}
			${'AnyAnotyherFont'}
			${'IconFont'}
			${null}
		`('[Font = "$font"]', ({ font }) => {
			const avatar = new ArrowComponent()
			document.body.appendChild(avatar)

			try {
				avatar.font = font

				it('Light DOM:', () => expect(avatar).toMatchSnapshot())
				it('Shadow DOM:', () =>
					expect(avatar.shadowRoot.children).toMatchSnapshot())
			} catch (error) {
				it('Error:', () => expect(error).toMatchSnapshot())
			}
		})
	})
})
