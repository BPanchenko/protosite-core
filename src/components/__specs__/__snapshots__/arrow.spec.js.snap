// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ArrowComponent] Attributes: [Font = "AnyAnotyherFont"] Error: 1`] = `[Error: Unsupported Font: "AnyAnotyherFont"]`;

exports[`[ArrowComponent] Attributes: [Font = "IconFont"] Light DOM: 1`] = `<c-arrow />`;

exports[`[ArrowComponent] Attributes: [Font = "IconFont"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    data-glyph="arrow"
    style="--icon-font: IconFont;"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Font = "null"] Light DOM: 1`] = `<c-arrow />`;

exports[`[ArrowComponent] Attributes: [Font = "null"] Light DOM: 2`] = `<c-arrow />`;

exports[`[ArrowComponent] Attributes: [Font = "null"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    data-glyph="arrow"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Font = "null"] Shadow DOM: 2`] = `
HTMLCollection [
  <i
    data-glyph="arrow"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-left-line"; Weight = "null"; Direction = "right"; Figure = "line"; Style = "oblique"] Light DOM: 1`] = `
<c-arrow
  direction="right"
  figure="line"
  glyph="arrow-left-line"
  style="oblique"
/>
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-left-line"; Weight = "null"; Direction = "right"; Figure = "line"; Style = "oblique"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    data-glyph="arrow-left-line"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-bottom-right-fill"; Size = "xxl"] Light DOM: 1`] = `
<c-arrow
  glyph="arrow-right-angle-bottom-right-fill"
/>
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-bottom-right-fill"; Size = "xxl"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    class="s-xxl"
    data-glyph="arrow-right-angle-bottom-right-fill"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-bottom-right-fill-angled"; Size = "sm"] Light DOM: 1`] = `
<c-arrow
  glyph="arrow-right-angle-bottom-right-fill-angled"
/>
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-bottom-right-fill-angled"; Size = "sm"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    class="s-sm"
    data-glyph="arrow-right-angle-bottom-right-fill-angled"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-right-bottom-fill"; Size = "lg"] Light DOM: 1`] = `
<c-arrow
  glyph="arrow-right-angle-right-bottom-fill"
/>
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-right-bottom-fill"; Size = "lg"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    class="s-lg"
    data-glyph="arrow-right-angle-right-bottom-fill"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-right-bottom-fill-acute"; Size = "another-size"] Error: 1`] = `[Error: Wrong size: "another-size"]`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-right-top-line"; Size = "null"] Error: 1`] = `[Error: Wrong size: "null"]`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-thick-bottom-left-fill-angled"; Size = "md"] Light DOM: 1`] = `
<c-arrow
  glyph="arrow-right-angle-thick-bottom-left-fill-angled"
/>
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-thick-bottom-left-fill-angled"; Size = "md"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    class="s-md"
    data-glyph="arrow-right-angle-thick-bottom-left-fill-angled"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-thick-right-top-fill-acute"; Size = "xs"] Light DOM: 1`] = `
<c-arrow
  glyph="arrow-right-angle-thick-right-top-fill-acute"
/>
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-thick-right-top-fill-acute"; Size = "xs"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    class="s-xs"
    data-glyph="arrow-right-angle-thick-right-top-fill-acute"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-thick-top-right-fill"; Size = "xl"] Light DOM: 1`] = `
<c-arrow
  glyph="arrow-right-angle-thick-top-right-fill"
/>
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-thick-top-right-fill"; Size = "xl"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    class="s-xl"
    data-glyph="arrow-right-angle-thick-top-right-fill"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-top-right-fill-acute"; Size = "xxs"] Light DOM: 1`] = `
<c-arrow
  glyph="arrow-right-angle-top-right-fill-acute"
/>
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-angle-top-right-fill-acute"; Size = "xxs"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    class="s-xxs"
    data-glyph="arrow-right-angle-top-right-fill-acute"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-line"; Size = "null"] Error: 1`] = `[Error: Wrong size: "null"]`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-line"; Weight = "null"; Direction = "null"; Figure = "null"; Style = "null"] Light DOM: 1`] = `
<c-arrow
  glyph="arrow-right-line"
/>
`;

exports[`[ArrowComponent] Attributes: [Glyph = "arrow-right-line"; Weight = "null"; Direction = "null"; Figure = "null"; Style = "null"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    data-glyph="arrow-right-line"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Glyph = "null"; Weight = "null"; Direction = "bottom-left"; Figure = "fill"; Style = "large-acute"] Light DOM: 1`] = `
<c-arrow
  direction="bottom-left"
  figure="fill"
  style="large-acute"
/>
`;

exports[`[ArrowComponent] Attributes: [Glyph = "null"; Weight = "null"; Direction = "bottom-left"; Figure = "fill"; Style = "large-acute"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    data-glyph="arrow-bottom-left-fill-large-acute"
  >
    <slot>
       
    </slot>
  </i>,
]
`;

exports[`[ArrowComponent] Attributes: [Glyph = "null"; Weight = "thick"; Direction = "bottom"; Figure = "null"; Style = "fill-angled"] Light DOM: 1`] = `
<c-arrow
  direction="bottom"
  style="fill-angled"
  weight="thick"
/>
`;

exports[`[ArrowComponent] Attributes: [Glyph = "null"; Weight = "thick"; Direction = "bottom"; Figure = "null"; Style = "fill-angled"] Shadow DOM: 1`] = `
HTMLCollection [
  <i
    data-glyph="arrow-thick-bottom-fill-angled"
  >
    <slot>
       
    </slot>
  </i>,
]
`;
